opencensus-migration/src/main/java/SpanCache.java

```diff
@@ -0,0 +1,70 @@
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import io.opencensus.implcore.trace.RecordEventsSpanImpl;
+import io.opencensus.trace.Span;
+import java.util.concurrent.TimeUnit;
+
+public class SpanCache {
+
+  private static final SpanCache SPAN_CACHE = new SpanCache();
+
+  private static final int MAXIMUM_CACHE_SIZE = 10000;
+  private static final int CACHE_EXPIRE_TIME = 10;
+  private static final TimeUnit CACHE_EXPIRE_UNIT = TimeUnit.MINUTES;
+
+  private static final CacheLoader<io.opentelemetry.trace.Span, Span> OT_TO_OC_CONVERTER =
+      new CacheLoader<io.opentelemetry.trace.Span, Span>() {
+        @Override
+        public Span load(io.opentelemetry.trace.Span span) {
+          return SpanConverter.fromOtelSpan(span);
+        }
+      };
+
+  private static final LoadingCache<io.opentelemetry.trace.Span, Span> OT_TO_OC =
+      CacheBuilder.newBuilder()
+          .maximumSize(MAXIMUM_CACHE_SIZE)
+          .expireAfterAccess(CACHE_EXPIRE_TIME, CACHE_EXPIRE_UNIT)
+          .build(OT_TO_OC_CONVERTER);
+
+  private static final CacheLoader<Span, io.opentelemetry.trace.Span> OC_TO_OT_CONVERTER =
+      new CacheLoader<Span, io.opentelemetry.trace.Span>() {
+        @Override
+        public io.opentelemetry.trace.Span load(Span span) {
+          return SpanConverter.toOtelSpan(span);
+        }
+      };
+
+  private static final LoadingCache<Span, io.opentelemetry.trace.Span> OC_TO_OT =
+      CacheBuilder.newBuilder()
+          .maximumSize(MAXIMUM_CACHE_SIZE)
+          .expireAfterAccess(CACHE_EXPIRE_TIME, CACHE_EXPIRE_UNIT)
+          .build(OC_TO_OT_CONVERTER);
+
+
+  public static SpanCache getInstance() {
+    return SPAN_CACHE;
+  }
+
+  private SpanCache() {}
+
+  io.opentelemetry.trace.Span toOtelSpan(Span ocSpan) {
+    io.opentelemetry.trace.Span otSpan = OC_TO_OT.getUnchecked(ocSpan);
+    OT_TO_OC.put(otSpan, ocSpan);
```

Comment:
This is not ideal.

Maybe we should instead replace two loading caches with just one to avoid this?
1. Define a key as a `traceId+spanId` string
2. Store OC and OT as a tuple.

```java
class SpanKey {
  // only one is set at a time
  Span ocSpan;
  io.opentelemetry.trace.Span otSpan;

  String key() {
    if (ocSpan != null) {
      // take traceId and spanId from OC span
    } else {
      // take traceId and spanId from OT span
    }
  }

  hashCode() {
    return key().hashCode();
  }
}

class SpanTuple {
  // both must be set
  Span ocSpan;
  io.opentelemetry.trace.Span otSpan;
}
```

```java
private static final LoadingCache<SpanKey, SpanTuple> CACHE = ...

private static final CacheLoader<SpanKey, SpanTuple> OC_TO_OT_CONVERTER =
      new CacheLoader<SpanKey, SpanTuple>() {
        @Override
        public SpanTuple load(SpanKey key) {
          Span ocSpan = key.getOCSpan();
          io.opentelemetry.trace.Span otSpan = key.getOTSpan();
          if (key.getOCSpan() != null) {
             return new SpanTuple(ocSpan, SpanConverter.toOtelSpan(ocSpan));
          }
          return return new SpanTuple(SpanConverter.fromOtelSpan(ocSpan), otSpan);
        }
      };
```

Let me know what you think.

Reply:
Alternatively, we may replace the LoadingCache with a non-loading one, as described in https://stackoverflow.com/a/13533148
and use `put` for both caches at once.
Then it will be clear that we add value to both caches at the same time.

Reply:
I've changed the cache to a non-loading one since it's easier to implement and understand

Reply:
Looks good, thanks!